File: lasso/lasso.R
library("glmnet")

parseClassifiable <- function(f) {
  # Parses a user-provided CSV file that is representative of a 
  # count matrix and returns its target vector and counts.
  #
  # Args:
  #   f: Input CSV file.
  #
  # Returns:
  #   x: Count matrix of dimensions i * j.
  #   y: Target vector of dimensions i * 1.
  
  df <- read.csv(file=f, header=T) # read matrix
  y <- as.matrix(df$Target) # extract target vector
  colnames(y) <- c('Target')
  seqs <- df$Sequence
  df$Length <- NULL # remove length vector
  df$Target <- NULL # remove target vector when extracted
  df$X <- NULL # an integer column solely for indexing
  df$Sequence <- NULL # first column not needed
  x <- as.matrix(df) # everything else are PWM counts
  rownames(x) <- seqs
  return(list('x'=x, 'y'=y)) # return matrix and target vector
}

buildLASSOClassifier <- function(x, y, nfold=3) {
  # Builds a LASSO classifier given a data matrix and target vector.
  #
  # Args:
  #   x: Count matrix of dimensions i * j.
  #   y: Target vector of dimensions i * 1.
  #   nfold: Number of cross-validation (CV) folds; default = 3.
  #
  # Returns:
  #   LASSO classifier.
  
  fit.cv <- cv.glmnet(x, y, nfolds=nfold, type.measure='auc', family='binomial')
  return(fit.cv)
}

getRatios <- function(x, y) {
  # Derives count-ratios per attribute within the count matrix. Resultant 
  # ratios capture attribute abundance within the target vector and serve
  # to shed light on attribute over-representation.
  #
  # Args:
  #   x: Count matrix of dimensions i * j
  #   y: Target vector of dimensions i * 1
  #
  # Returns:
  #   Matrix containing attribute ratios and count enumerations.
  
  stopifnot(length(table(y)) == 2) # y vector must have 0 or 1 values only 
  sum.query <- colSums(x[which(y == 1), ]) # sum rows referencing query targets
  sum.control <- colSums(x[which(y == 0), ]) # repeat for control
  ratio <- round(sum.query / sum.control, 2) # derive ratio
  m <- as.matrix(cbind(ratio, sum.query, sum.control)) # gather results in matrix
  m <- as.matrix(m[order(row.names(m)), ]) # order by row-name
  colnames(m) <- c('Ratio', 'Num.Query', 'Num.Control')
  return(m)
}

getPValues <- function(x, y, adj.method="BH") {
  # Derives adjusted p-values for each count matrix attribute based on 
  # abundance of the observation.
  #
  # Args:
  #   x: Count matrix of dimensions i * j
  #   y: Target vector of dimensions i * 1
  #   method: P-value adjustment method. See p.adjust(method,...)
  #
  # Returns:
  #   P-value (adjusted) vector.
  
  ratios <- getRatios(x, y) # contains feature counts
  total.query <- sum(ratios[, 2]) # sum query column
  total.control <- sum(ratios[, 3]) # sum control column
  p.vals <- matrix(ncol=1, nrow=nrow(ratios)) # for storing p-values
  rownames(p.vals) <- rownames(ratios)

  for (i in 1: nrow(ratios)) { # iteratively compute matrix p-value
    val.query <- ratios[i, 2]
    val.control <- ratios[i, 3]
    m <- matrix(c(val.query, val.control, total.query-val.query, total.control-val.control),
                ncol=2, byrow=T)
    p.vals[i, 1] <- fisher.test(m)$p.value # save p-value
  }
  p.vals[, 1] <- p.adjust(p.vals, method=adj.method) # adjust p-values
  colnames(p.vals) <- c('P.value')
  return(p.vals)
}

toPredictionVector <- function(x, y, iter=1, nfold=5) {
  # A prediction matrix is the result of iteratively (and randomly) building 
  # classifiers and deriving predictions of each observation.
  # Upon completion, each observation references predictions derived per
  # iteration, useful in possible dimensionality reduction.
  #
  # Args:
  #   x: Count matrix of dimensions i * j.
  #   y: Target vector of dimensions i * 1.
  #   iter: Number of iterations to perform.
  #   nfold: Number of cross-validations to perform.
  #
  # Returns:
  #   Vector of predictions of dimensions i * 1. 
  
  stopifnot(length(table(y)) == 2) # y vector must have 0 or 1 values only 
  matrix.control <- x[which(y == 0), ]
  matrix.query <- x[which(y == 1), ]
  perc.train <- (nfold -1) / nfold # fraction as-to how much of matrix is for testing
  list.aucs <- rep(0, iter) # generate vector of AUCs representative of each iteration
  all.preds <- matrix(nrow=nrow(x), ncol=iter) # create matrix of all LASSO predictions
  rownames(all.preds) <- rownames(x) # save row names
  
  for (i in 1: iter) {
    time.start <- Sys.time() # get start time
    query.train <-   matrix.query[sample.int(nrow(matrix.query), size=floor(nrow(matrix.query) * perc.train)), ]
    control.train <- matrix.control[sample.int(nrow(matrix.control), size=floor(nrow(matrix.control) * perc.train)), ]
    sample.x <- rbind(query.train, control.train) # join two matrices
    sample.y <- as.matrix(c(rep(1, nrow(query.train)),  
                            rep(0, nrow(control.train)))) # add target vector
    cat("[ Iteration ", i, '/', iter, ' ] ... ')
    sample.fit.cv <- buildLASSOClassifier(x=sample.x, y=sample.y, nfold=nfold)
    iter.preds <- doPrediction(sample.fit.cv, x) # derive predictions
    all.preds[, i] <- iter.preds # save predictions produced from iteration to column i
    list.aucs[i] <- getAUC(sample.fit.cv) # set the desired AUC per iteration    
    time.iter <- format(.POSIXct(difftime(Sys.time(), time.start, units=c("secs")),tz="GMT"), "%H:%M:%S")
    cat(time.iter, 'h:m:s \n') # print on existing line.
  }
  all.preds <- rowSums(all.preds) / iter # average predictions for each observation
  return(list('preds'=all.preds, 'aucs'=list.aucs)) # return observation-specific predictions
}

generateReport <- function(weights, ratios, out='./report.csv') {
  # Generates a report given a LASSO fit, PWM ratios, and their p-values.
  #
  # Args:
  #   weights: LASSO weights derived using getWeights.
  #   ratios: Feature ratios derived using getRatios
  #   out: Output file to save results to.
  #
  # Returns:
  #   Output results file defined by the out argument.
  
  imp <- weights * ratios[, 1] # importance => weight * ratio
  colnames(imp) <- c('Importance')
  report.mat <- cbind(weights, ratios, imp)

  # test that row-names are shared between weights and ratios
  stopifnot(identical(rownames(weights), rownames(ratios)))
  write.csv(report.mat, out) # write merged martix to a file
}

homogenize <- function(x, y, preds, threshold = 0.5) {
  # Given a count matrix and a target vector, each row is enumerated against 
  # to determine if its predictions cumulatively exceed a user-provided threshold.
  # For query data-points to exceed, their cumulative prediction must exceed the
  # threshold. Query data-points not exceeding this threshold and removed along
  # with their corresponding control.
  # Implementation is based on Narlikar et. al., Genome Research, 2010.
  #
  # Args:
  #   x: Count matrix of dimensions i * j
  #   y: Target vector of dimensions i * 1
  #   preds:  Matrix of observation predictions of dimensions i * k
  #   threshold: prediction cutoff.
  #
  # Returns:
  #   List referencing homogenized count matrix and target vector.
  
  control.counts <- x[which(y == 0), ] # get matrix representing control and query data
  query.counts <- x[which(y == 1), ]
  query.preds <- preds[which(y == 1)] #>= threshold # uninterested in control predictions.
  
  # iterate over each query and check if it has a corresponding control; a
  # test purely determined if the query substring is within the control.
  new.query <- c() # create vector so rows can be added without matrix usage.
  new.control <- c() 
  names.query <-c() # references row-names as it is unknown which observations are/aren't valid.
  names.control <- c()
  rownames.control <- row.names(control.counts)
  
  for (i in 1: nrow(query.counts)) {
    query.name <- row.names(query.counts)[i]
    query.pred <- query.preds[i] # whether the query passes threshold
    match.pos <- grep(query.name, rownames.control, fixed=T) # vector of indices

    # if query prediction is true, add control and query data-points.
    if (query.pred >= threshold) {
      new.query <- rbind(new.query, query.counts[i, ]) # add passed queries.
      names.query <- rbind(names.query, query.name) # add query name
      
      # in the case of small datasets, a query may not have a corresponding
      # control, thus handle instances whereby no match is found and add if so.
      if (length(match.pos) > 0) { # if query maps to control, get first hit.
        control.name <- rownames.control[match.pos[1]] # fetch first hit only
        #cat( '   ',control.name, '\n')
        new.control <- rbind(new.control, control.counts[match.pos, ]) # add its respective control.
        names.control <- rbind(names.control, control.name) # add control name
      }
    }
  }
  
  rownames(new.query) <- names.query # add row-names to each data-frame.
  rownames(new.control) <- names.control
  new.x <- rbind(new.query, new.control) # merge counts into new matrix
  new.y <- as.matrix(c(rep(1, nrow(new.query)), # build target vector
                       rep(0, nrow(new.control))))
  return(list('x'=new.x, 'y'=new.y))
}

saveJob <- function(proj.name) {
  # Saves the current workspace and corresponding history.
  #
  # Args:
  #   proj.name: Project name to save workspace and history as.
  
  save.image(file=paste(proj.name, '.RData', sep=''))
  savehistory(file=paste(proj.name,'.RHistory', sep=''))
  cat(paste('Workspace, history saved to', getwd(), '\n' ,sep=' ')) 
}

doPrediction <- function(fit.cv, x) {
  # Performs predictions given a count-matrix and a corresponding classifier.
  # Args:
  #   fit.cv: LASSO cross-validated (CV) classifier.
  #   x: Matrix of dimensions i * j.
  #
  # Returns:
  #   preds: vector of size i * 1 representing predictions per matrix row.
  
  preds <- predict(fit.cv, x, type='response', s=c('lambda.min')) # derive predictions
  return(preds)
}

getWeights <- function(fit.cv, scale.weights=T) {
  # Yields attribute weights (coefficients) given LASSO classifier.
  # Weights can be scaled based on whether they are positive or
  # negative. Such scaling is based off Taher, et. al., 2013.
  #
  # Args:
  #   fit.cv: LASSO cross-validated (CV) classifier.
  #   scale.weights: Z-normalize LASSO weights.
  #
  # Returns:
  #   weights: Dataframe of attributes and their respective weights.

  m <- as.matrix(coef(fit.cv, s=min(fit.cv$lambda)))
  if (scale.weights) { # standardize LASSO weights, if sought.
    weights.min <- min(m)
    weights.max <- max(m)
    for (i in 1: length(m)) {
      val <- m[i, 1] # retrieve the current value
      if (val < 0) { # normalize weights less than zero
        m[i, 1] <- -1 * ( 1 - ( ( val - weights.min ) / -weights.min ) )
      }
      else { # normalize weights greater-than or equal to zero
        m[i, 1] <- val / weights.max # maximum weight is set to 1.0
      }
    }
  }
  
  m <- as.matrix(m[order(row.names(m)), ]) # order weights by row-name
  m <- as.matrix(m[row.names(m) != '(Intercept)', ]) # intercept weight not needed
  colnames(m) <- c('Weight')
  return(m)
}

getAUC <- function(fit.cv) {
  # Identifies the Area Under Curve (AUC) of a LASSO classifier.
  #
  # Args:
  #   fit.cv: LASSO cross-validated (CV) classifier.
  #
  # Returns:
  #   auc: Classifier AUC.
  
  min.lambda <- log(fit.cv$lambda.min) # lambda of maximal AUC; i.e. minimal CV error
  pos.auc <- which(log(fit.cv$lambda) == log(fit.cv$lambda.min), arr.ind=T) # get min lambda position
  auc <- fit.cv$cvm[pos.auc]
  return(auc)
}

File: lasso/lasso.py
'''
'''

import argparse
import pandas
import sklearn


class MatrixParser():
    '''
    The MatrixParser class provides behaviors and states for parsing a matrix
    CSV file. Such files must have their first column represent desired 
    observations. The very last column must explicitly represent a binary
    target vector (0, 1).
    '''
    COLNAME_PWM = 'Unnamed: 0'
    COLNAME_TARGET = 'Target'

    def __init__(self, f, is_classifiable=True):
        '''
        Creates an object of type MatrixParser. Each instance of this object
        enables the ability to parse a matrix saved as a CSV file and mold its
        contents to ultimately be classified.
        @param f: Input CSV file.
        @param is_classifiable: Whether has file has a target vector or not.
        '''
        self.mat = pandas.read_csv(f)
        self.is_classifiable = is_classifiable
        self.sanity_check()

    def sanity_check(self):
        '''
        Perform sanity checks to ensure that the first and last columns of a
        user-provided CSV file reference PWMs and Targets (if applicable).
        The former check is only applicable if a classifiable matrix is
        provided. Ideally, such a file should have an unnamed first column
        which references PWMs, and an explicitly-named target column that is
        the very last column in the file.
        '''
        print('Sanity-checking matrix ...')
        cols = list(self.mat.columns)
        has_pwm = cols[0] == MatrixParser.COLNAME_PWM  # PWMs must be found.
        has_target = cols[-1] == MatrixParser.COLNAME_TARGET
        if self.is_classifiable:  # targets, PWMs needed in classifiable files.
            if not all([has_pwm, has_target]):
                raise IOError('PWMs, Targets must be first and last columns.')
            if set(self.mat[MatrixParser.COLNAME_TARGET]) != set([0, 1]):
                raise IOError('Target vector must be binary (0, 1) only.')
        else:
            if not has_pwm:
                raise IOError('PWMs must be the first column.')

    def parse(self):
        '''
        Parse the CSV file. Following such parsing, the matrix is accessible
        through its column-names and row-names.
        '''
        print('Parsing matrix ...')
        self.mat = self.mat.set_index(MatrixParser.COLNAME_PWM, drop=True)
        self.mat.index.names = ['PWM']
        if 'Sequence' in self.mat.columns:  # drop column already indexed.
            self.mat = self.mat.drop(['Sequence'], axis=1)

    def to_classifiable_matrix(self):
        '''
        Wraps the results following matrix parsing as an object of type
        ClassifiableMatrix. An object of this class is therefore fit for
        matrix manipulation as it is now fit to perform classification on.
        @return: object of type ClassifiableMatrix.
        '''
        cm = ClassifiableMatrix(self.mat)
        if cm.nrows() != cm.get_query().nrows() + cm.get_control().nrows():
            raise IOError('Error in matrix; ensure target is binary.')
        cm.debug()
        return cm


class ClassifiableMatrix():
    '''
    A ClassifiableMatrix enables the manipulation of a matrix data-structure
    with intention of classification. Such an object must have observations
    as designated by row-names, and features as designated by column names.
    A corresponding binary target vector need also be present; found as the
    last column in the matrix.
    '''
    def __init__(self, m):
        '''
        Creates an object of type ClassifiableMatrix.
        @param m: pandas Dataframe object generated from MatrixParser.
        '''
        assert isinstance(m, pandas.DataFrame)
        self.mat = m

    def debug(self):
        n_query, n_cont = self.get_query().nrows(), self.get_control().nrows()
        print('[ {0} rows x {1} columns ]'.format(self.nrows(), self.ncols()))
        print('[ => {0} query, {1} control ]'.format(n_query, n_cont))

    def normalize(self, n):
        '''
        In cases whereby a query-set is small, a corresponding control-set
        cannot be of equal size since this control-set may be capture the the
        properties of the global system. Thus, rather than generating 1 control
        for each query, you may have to generate N controls. Doing so could
        help capture the systematic properties of the query by giving it a
        large control-set to be contrasted against. This constant, N, is what
        is used to 'amplify' counts and must be normalized so as to not skew
        classifications.
        '''
        control = self.mat[MatrixParser.COLNAME_TARGET] == 0  # get controls
        self.mat.loc[control] = self.mat.loc[control] / n  # normalize

    def get_control(self):
        '''
        Retrieve the background observations which are observations referencing
        a target value of 0.
        @return: ClassifiableMatrix object referencing control observations.
        '''
        control = self.mat[self.mat[MatrixParser.COLNAME_TARGET] == 0]
        return ClassifiableMatrix(control)

    def get_query(self):
        '''
        Retrieve the signal (query) observations which are observations
        referencing a target value of 1.
        @return: ClassifiableMatrix object referencing query observations.
        '''
        query = self.mat[self.mat[MatrixParser.COLNAME_TARGET] == 1]
        return ClassifiableMatrix(query)

    def nrows(self):
        '''
        Retrieve the total number of rows.
        @return: integer row count.
        '''
        return self.mat.shape[0]

    def ncols(self):
        '''
        Retrieve the total number of columns.
        @return: integer column count.
        '''
        return self.mat.shape[1]


class ClassificationFactory():
    '''
    The ClassificationFactory class enables the ability to classify a
    ClassifiableMatrix object. Successful classification is indeed a much
    desired objective and can be obtainable with the help of numerous
    runtime parameters such as cross-validation.
    '''
    def __init__(self, m, iters, cv, thr, homogen):
        '''
        Creates a ClassificationFactory object, an object designed solely for
        classifying features based on the observation target vector.
        @param m: ClassificationMatrix object.
        @param iters: Number of predictive iterations.
        @param cv: Number of cross-validations.
        @param thr: Predictive threshold cutoff.
        @param homogen: Whether to homogenize counts (see Taher, et. al, 2013).
        '''
        self.m = m
        self.iters = iters
        self.cv = cv
        self.thr = thr
        self.homogenize = homogen


def main(args):
    if args['in']:
        m = MatrixParser(f=args['in'])
        m.parse()
        cm = m.to_classifiable_matrix()
        cm.normalize(n=args['amplify'])  # normalize controls if amplified.
        cf = ClassificationFactory(m=cm, iters=args['iters'], cv=args['cv'],
                                   thr=args['thr'], homogen=args['homogenize'])

if __name__ == '__main__':
    desc = 'LASSO classification and prediction.'
    parser = argparse.ArgumentParser(description=desc, add_help=False)
    parser_class = parser.add_argument_group('Classification')
    parser_pred = parser.add_argument_group('Prediction')
    parser_misc = parser.add_argument_group('Miscellaneous')

    # Define classification parameters
    parser_class.add_argument('-in', metavar='CSV',
                              help='Count-matrix with binary targets [na]')
    parser_class.add_argument('-iters', metavar='INT', default=3, type=int,
                              help='# / homogenization iterations [3]')
    parser_class.add_argument('-cv', metavar='INT', default=5, type=int,
                              help='# / cross-validations [5]')
    parser_class.add_argument('-thr', metavar='FLOAT', default=0.5, type=float,
                              help='Prediction probability threshold [0.5]')
    parser_class.add_argument('-amplify', metavar='INT', default=1, type=int,
                              help='Divide pre-amplified control counts [1]')
    parser_class.add_argument('--homogenize', default=False, metavar='',
                              help='Perform homogenization [false]')

    # Define prediction-specific parameters
    parser_pred.add_argument('-counts', metavar='CSV',
                             help='Count matrix for deriving predictions [na]')
    parser_pred.add_argument('-model', metavar='MODEL',
                             help='Classifier object (.model) [na]')

    # Define miscellaneous parameters
    parser_misc.add_argument('-out', metavar='DIR', default='./proj',
                             help='Output folder [./proj]')
    parser_misc.add_argument('-h', action='help',
                             help='Show this help message and exit.')

    args = vars(parser.parse_args())
    try:
        main(args)
    except IOError as e:
        print(e)
    except OSError as e:
        print(e)
